### PR TITLE
feat: use `SoftFailure` in `Weeder2` at `pickRestrictableChoosePattern` (Error Var)

### PR DESCRIPTION
**Pull request by jaschdoc**
Related to https://github.com/flix/flix/pull/8586
Related to https://github.com/flix/flix/issues/8540

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 1]
```diff
@@ -2551,7 +2559,7 @@ object Weeder2 {
       }
     }
 
-    private def tryPickLatticeTermExpr(tree: Tree): Validation[Option[Expr], CompilationMessage] = {
+    private def tryPickLatticeTermExpr(tree: Tree)(implicit flix: Flix): Validation[Option[Expr], CompilationMessage] = {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:2562]: implicit flix is no longer used, right?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:2562]: @jaschdoc What do you say?

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 2]
```diff
@@ -1389,16 +1389,24 @@ object Weeder2 {
...
+                  Validation.toSoftFailure(WeededAst.RestrictableChoosePattern.Error( other.loc.asSynthetic), e)
               }
-            case other => Validation.toHardFailure(UnsupportedRestrictedChoicePattern(isStar, other.loc))
+            case other =>
+              val e = UnsupportedRestrictedChoicePattern(isStar, other.loc)
+              Validation.toSoftFailure(List(WeededAst.RestrictableChoosePattern.Error(other.loc.asSynthetic)), e)
           }
           mapN(inner)(RestrictableChoosePattern.Tag(qname, _, loc))
-        case other => Validation.toHardFailure(UnsupportedRestrictedChoicePattern(isStar, other.loc))
+        case other =>
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:1401]: @mlutze here

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala (hunk 1 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:2562]: Yes, it's used in the `other` case of `pickRestrictableChoosePattern`

**Comment by jaschdoc** [comment]
If we combine this PR with https://github.com/flix/flix/pull/8586 we get 100% error nodes instead of some `Var` error nodes and and on fake `Tag` node

**Comment by jaschdoc** [comment]
I think we should merge this one first, though, because it has the minimal amount of changes

File: main/src/ca/uwaterloo/flix/language/ast/KindedAst.scala [hunk 3]
```diff
@@ -270,6 +270,8 @@ object KindedAst {
 
     case class Var(sym: Symbol.VarSym, tvar: Type.Var, loc: SourceLocation) extends VarOrWild
 
+    case class Error(tvar: Type.Var, loc: SourceLocation) extends VarOrWild
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/KindedAst.scala:273]: Error should always be the last

**Comment by magnus-madsen** [comment]
So what should I do? Give feedback on this one?

**Comment by jaschdoc** [comment]
> So what should I do? Give feedback on this one?

Yes, and if we merge this then we deal with https://github.com/flix/flix/pull/8586

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala (hunk 2 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:1400]: I'm confused. Why do we construct a tag?

File: main/src/ca/uwaterloo/flix/language/ast/DesugaredAst.scala [hunk 4]
```diff
@@ -271,6 +271,8 @@ object DesugaredAst {
 
     case class Var(ident: Name.Ident, loc: SourceLocation) extends VarOrWild
 
+    case class Error(loc: SourceLocation) extends VarOrWild
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/DesugaredAst.scala:274]: Error should be last

File: main/src/ca/uwaterloo/flix/language/ast/NamedAst.scala [hunk 5]
```diff
@@ -280,6 +280,8 @@ object NamedAst {
 
     case class Var(sym: Symbol.VarSym, loc: SourceLocation) extends VarOrWild
 
+    case class Error(loc: SourceLocation) extends VarOrWild
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/NamedAst.scala:283]: Error should be last

File: main/src/ca/uwaterloo/flix/language/ast/ResolvedAst.scala [hunk 6]
```diff
@@ -286,6 +286,8 @@ object ResolvedAst {
 
     case class Var(sym: Symbol.VarSym, loc: SourceLocation) extends VarOrWild
 
+    case class Error(loc: SourceLocation) extends VarOrWild
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ResolvedAst.scala:289]: Error should be last

File: main/src/ca/uwaterloo/flix/language/ast/WeededAst.scala [hunk 7]
```diff
@@ -297,6 +297,8 @@ object WeededAst {
 
     case class Var(ident: Name.Ident, loc: SourceLocation) extends VarOrWild
 
+    case class Error(loc: SourceLocation) extends VarOrWild
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/WeededAst.scala:300]: Error should be last

**Comment by magnus-madsen** [comment]
Added some comments. See also Matt's comment.

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala (hunk 2 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:1400]: Before this we returned a `HardFailure` but that requires `UnsupportedRestrictedChoicePattern` to be `Unrecoverable`. Now that we introduce `SoftFailure`s we require that the error i `Recoverable`. This means we cannot return the `HardFailure` like before so we are forced to return a value. Returning an error node i what https://github.com/flix/flix/pull/8586 does. Note however, that this returns a `Tag` with and `Error` inside. I think we should combine the two PRs (by first merging this).

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:1400]: Ah, so you'll undo all the implicit flixes?

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:1400]: Yes :)

**Comment by magnus-madsen** [comment]
Conflicted. Lets resolve and then @mlutze can merge.

File: main/src/ca/uwaterloo/flix/language/ast/WeededAst.scala [hunk 8]
```diff
@@ -299,6 +299,8 @@ object WeededAst {
 
     case class Tag(qname: Name.QName, pat: List[VarOrWild], loc: SourceLocation) extends RestrictableChoosePattern
 
+    case class Error(loc: SourceLocation) extends VarOrWild
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/ast/WeededAst.scala:302]: @magnus-madsen This has to be both `VarOrWild` and `RestrictableChoosePattern` which is the core of the issue

**Review by magnus-madsen — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/ast/WeededAst.scala (hunk 8 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/WeededAst.scala:302]: OK 👍

**Pull request merged**